src/sinks/influxdb/mod.rs

```diff
@@ -0,0 +1,778 @@
+        v1_settings,
+        v2_settings
+    ))]
+    BothConfiguration {
+        v1_settings: InfluxDB1Settings,
+        v2_settings: InfluxDB2Settings,
+    },
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct InfluxDB1Settings {
+    database: String,
+    consistency: Option<String>,
+    retention_policy_name: Option<String>,
+    username: Option<String>,
+    password: Option<String>,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+pub struct InfluxDB2Settings {
+    org: String,
+    bucket: String,
+    token: String,
+}
+
+trait InfluxDBSettings {
+    fn write_uri(self: &Self, endpoint: String) -> crate::Result<Uri>;
+    fn healthcheck_uri(self: &Self, endpoint: String) -> crate::Result<Uri>;
+    fn token(self: &Self) -> String;
+}
+
+impl InfluxDBSettings for InfluxDB1Settings {
+    fn write_uri(self: &Self, endpoint: String) -> crate::Result<Uri> {
+        encode_uri(
+            &endpoint,
+            "write",
+            &mut [
+                ("consistency", self.consistency.clone()),
+                ("db", Some(self.database.clone())),
+                ("rp", self.retention_policy_name.clone()),
+                ("p", self.password.clone()),
+                ("u", self.username.clone()),
+                ("precision", Some("ns".to_owned())),
+            ],
+        )
+    }
+
+    fn healthcheck_uri(self: &Self, endpoint: String) -> crate::Result<Uri> {
+        encode_uri(&endpoint, "ping", &mut [])
+    }
+
+    fn token(self: &Self) -> String {
+        "".to_string()
+    }
+}
+
+impl InfluxDBSettings for InfluxDB2Settings {
+    fn write_uri(self: &Self, endpoint: String) -> crate::Result<Uri> {
+        encode_uri(
+            &endpoint,
+            "api/v2/write",
+            &mut [
+                ("org", Some(self.org.clone())),
+                ("bucket", Some(self.bucket.clone())),
+                ("precision", Some("ns".to_owned())),
+            ],
+        )
+    }
+
+    fn healthcheck_uri(self: &Self, endpoint: String) -> crate::Result<Uri> {
+        encode_uri(&endpoint, "health", &mut [])
+    }
+
+    fn token(self: &Self) -> String {
+        self.token.clone()
+    }
+}
+
+fn influxdb_settings(
+    influxdb1_settings: Option<InfluxDB1Settings>,
+    influxdb2_settings: Option<InfluxDB2Settings>,
+) -> Result<Box<dyn InfluxDBSettings>, crate::Error> {
+    if influxdb1_settings.is_some() & influxdb2_settings.is_some() {
+        return Err(ConfigError::BothConfiguration {
+            v1_settings: influxdb1_settings.unwrap(),
+            v2_settings: influxdb2_settings.unwrap(),
+        }
+        .into());
+    }
+
+    if influxdb1_settings.is_none() & influxdb2_settings.is_none() {
+        return Err(ConfigError::MissingConfiguration {}.into());
+    }
+
+    if let Some(settings) = influxdb1_settings {
+        Ok(Box::new(settings))
+    } else {
+        Ok(Box::new(influxdb2_settings.unwrap()))
+    }
```

Comment:
This whole block can be replaced with a much simpler `match`:
```suggestion
    match (influxdb1_settings, influxdb2_settings) {
        (None, None) => Err(ConfigError::MissingConfiguration.into()),
        (Some(v1_settings), Some(v2_settings)) => Err(ConfigError::BothConfiguration { v1_settings, v2_settings }.into()),
        (Some(settings), None) => Ok(Box::new(settings)),
        (None, Some(settings)) => Ok(Box::new(settings)),
    }
```
(`rustfmt` will probably re-indent this block)